src/xdp-utils.c

```diff
@@ -1190,3 +1196,432 @@ xdp_has_path_prefix (const char *str,
+  parser = json_parser_new ();
+  if (!json_parser_load_from_data (parser, data, len, error))
+    {
+      g_prefix_error (error, "Could not parse '%s': ", path);
+      return 0;
+    }
+
+  root = json_parser_get_root (parser);
+  if (!root)
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+		   "Could not parse '%s': empty file", path);
+      return 0;
+    }
+
+  if (!JSON_NODE_HOLDS_OBJECT (root))
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+		   "Could not parse '%s': invalid structure", path);
+      return 0;
+    }
+
+  cpo = json_node_get_object (root);
+
+  pid = json_object_get_int_member (cpo, "child-pid");
+  if (pid == 0)
+    g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+		 "Could not parse '%s': child-pid missing", path);
+
+  return pid;
+}
+
+#define xdp_lockguard G_GNUC_UNUSED __attribute__((cleanup(xdp_auto_unlock_helper)))
+
+static gboolean
+xdg_app_info_ensure_pidns (XdpAppInfo  *app_info,
+			   DIR         *proc,
+			   GError     **error)
+{
+  xdp_lockguard GMutex *guard = NULL;
+  xdp_autofd int fd = -1;
+  pid_t pid;
+  ino_t ns;
+  int r;
+
+  guard = xdp_auto_lock_helper (app_info->u.flatpak.pidns_lock);
+
+  if (app_info->u.flatpak.pidns_id != 0)
+    return TRUE;
+
+  pid = xdp_app_info_get_child_pid (app_info, error);
+  if (pid == 0)
+    return FALSE;
+
+  fd = open_pid_fd (dirfd (proc), pid, error);
+  if (fd == -1)
+    return FALSE;
+
+  r = lookup_ns_from_pid_fd (fd, &ns);
+  if (r < 0)
+    {
+      int code = g_io_error_from_errno (-r);
+      g_set_error (error, G_IO_ERROR, code,
+		   "Could not query /proc/%u/ns/pid: %s",
+		   (guint) pid, g_strerror (-r));
+      return FALSE;
+    }
+
+  app_info->u.flatpak.pidns_id = ns;
+
+  return TRUE;
+}
+
+gboolean
+xdg_app_info_map_pids (XdpAppInfo  *app_info,
+		       pid_t       *pids,
+		       guint        n_pids,
+		       GError     **error)
+{
+  gboolean ok;
+  DIR *proc;
+  uid_t uid;
+  ino_t ns;
+
+  g_return_val_if_fail (app_info != NULL, FALSE);
+  g_return_val_if_fail (pids != NULL, FALSE);
+
+  if (app_info->kind != XDP_APP_INFO_KIND_FLATPAK)
+    {
+      g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
+			   "Mapping pids is not supported.");
+      return FALSE;
+    }
+
+  proc = opendir ("/proc");
+  if (proc == NULL)
+    {
+      g_set_error (error, G_IO_ERROR, g_io_error_from_errno (errno),
+		   "Could not open '/proc: %s", g_strerror (errno));
+      return FALSE;
+    }
+
+  /* Make sure we know the pid namespace the app is running in */
+  ok = xdg_app_info_ensure_pidns (app_info, proc, error);
+  if (!ok)
+    {
+      g_prefix_error (error, "Could not determine pid namespace: ");
+      goto out;
+    }
+
+  /* we also make sure the real user id matches
+   * to the process owner we are trying to resolve
+   */
+  uid = getuid ();
+
+  ns = app_info->u.flatpak.pidns_id;
+  ok = map_pids (proc, ns, pids, n_pids, uid, error);
+
+ out:
+  closedir (proc);
+  return ok;
```

Comment:
This looks like quite a bit of work, and ugly and racy. There is no better way to get this information from the kernel ?

Reply:
Currently not, sadly. There is a plan for having a API to translate pids, but it is uncertain when it will land; I think 5.3 would be the earliest. Once we have support, I will add it and have the existing code only as fallback.